istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -222,23 +227,92 @@ func createRemoteSecretFromTokenAndServer(tokenSecret *v1.Secret, clusterName, s
 	return createRemoteServiceAccountSecret(kubeconfig, clusterName, secName)
 }
 
-func getServiceAccountSecret(kube kubernetes.Interface, saName, saNamespace string) (*v1.Secret, error) {
-	serviceAccount, err := kube.CoreV1().ServiceAccounts(saNamespace).Get(context.TODO(), saName, metav1.GetOptions{})
+func getServiceAccountSecret(kube kubernetes.Interface, opt RemoteSecretOptions) (*v1.Secret, error) {
+	// Create the service account if it doesn't exist.
+	serviceAccount, err := getOrCreateServiceAccount(kube, opt)
 	if err != nil {
 		return nil, err
 	}
+
 	if len(serviceAccount.Secrets) != 1 {
 		return nil, fmt.Errorf("wrong number of secrets (%v) in serviceaccount %s/%s",
-			len(serviceAccount.Secrets), saNamespace, saName)
+			len(serviceAccount.Secrets), opt.Namespace, opt.ServiceAccountName)
 	}
+
 	secretName := serviceAccount.Secrets[0].Name
 	secretNamespace := serviceAccount.Secrets[0].Namespace
 	if secretNamespace == "" {
-		secretNamespace = saNamespace
+		secretNamespace = opt.Namespace
 	}
 	return kube.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
 }
 
+func getOrCreateServiceAccount(kube kubernetes.Interface, opt RemoteSecretOptions) (*v1.ServiceAccount, error) {
+	allSAs, err := kube.CoreV1().ServiceAccounts(opt.Namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, sa := range allSAs.Items {
+		if sa.Name == opt.ServiceAccountName {
+			// Found it.
+			return &sa, nil
+		}
+	}
+
+	// Create a basic IOP File to install the minimum configuration for the service account.
+	iopFile, err := createServiceAccountIOPFile()
+	if err != nil {
+		return nil, err
+	}
+
+	var setOverlay []string
+	if len(opt.ManifestsPath) > 0 {
+		setOverlay = []string{"installPackagePath=" + opt.ManifestsPath}
+	}
+
+	// Install the configuration to create the service account.
+	if err := mesh.InstallManifests(setOverlay, []string{iopFile}, false, false,
+		opt.KubeOptions.Kubeconfig, opt.KubeOptions.Context, 30*time.Second, opt.Logger); err != nil {
+		return nil, err
+	}
+
+	// Return the newly created service account.
+	return kube.CoreV1().ServiceAccounts(opt.Namespace).Get(
+		context.TODO(), opt.ServiceAccountName, metav1.GetOptions{})
+}
+
+func createServiceAccountIOPFile() (string, error) {
+	iopFile, err := ioutil.TempFile("", "remote-secret-iop-*")
+	if err != nil {
+		return "", err
+	}
+	// Close the file after we're done.
+	defer func() { _ = iopFile.Close() }()
+
+	// Generate the IOP file.
+	iopContent := `
+apiVersion: install.istio.io/v1alpha1
+kind: IstioOperator
+spec:
+  profile: empty
+  components:
+    base:
+      enabled: true
```

Comment:
This will install all the crds and other service accounts from `base` as well, right?  Seems it is more than you want ( just create the istio-reader-service-account sa)

Reply:
We also want the ClusterRole and other pieces.

Reply:
We probably want these to come from the templates/manifests some way or another.

Reply:
@irisdingbj do you have a suggestion for how to strip down the installation further? AFAIK this is the most lean installation we support currently?

@howardjohn @stevenctl any thoughts?

Reply:
FYI I've updated the config to not install crds. I think that's about all we can do. There isn't much else in it, really.

Reply:
We have a `valuex.base.enableCRDTemplates` which could filter some things out. 

There are quite a few other things in base that require template values for revision, system namespace, enableAnalysis as well as setting up the `remotePilotAddress` endpoint (if specified). 

Practically, it shouldn't really matter that we add these extra things because they'll just be replaced when we do the full install that includes ingress.

Reply:
```
CustomResourceDefinition/authorizationpolicies.security.istio.io.
CustomResourceDefinition/destinationrules.networking.istio.io.
CustomResourceDefinition/envoyfilters.networking.istio.io.
CustomResourceDefinition/gateways.networking.istio.io.
CustomResourceDefinition/istiooperators.install.istio.io.
CustomResourceDefinition/peerauthentications.security.istio.io.
CustomResourceDefinition/requestauthentications.security.istio.io.
CustomResourceDefinition/serviceentries.networking.istio.io.
CustomResourceDefinition/sidecars.networking.istio.io.
CustomResourceDefinition/virtualservices.networking.istio.io.
CustomResourceDefinition/workloadentries.networking.istio.io.
CustomResourceDefinition/workloadgroups.networking.istio.io.
ServiceAccount/istio-reader-service-account.istio-system
ServiceAccount/istiod-service-account.istio-system
ClusterRole/istio-reader-istio-system.
ClusterRole/istiod-istio-system.
ClusterRoleBinding/istio-reader-istio-system.
ClusterRoleBinding/istiod-istio-system.
Role/istiod-istio-system.istio-system
RoleBinding/istiod-istio-system.istio-system
```
is what is installed which seems ore than expected.

Do we we just want
```
ServiceAccount/istio-reader-service-account.istio-system
ClusterRole/istio-reader-istio-system.
ClusterRoleBinding/istio-reader-istio-system.
```
?

Reply:
Yes that's basically what we want. We definitely don't want the CRDs, but it doesn't seem currently possible today (the `Values.base.enableCRDTemplates` doesn't seem to have any effect.

FYI @irisdingbj @linsun

Reply:
created https://github.com/istio/istio/issues/28087 to track.

Reply:
`Values.base.enableCRDTemplate` is used for helm2 support I believe when we moved crds under a separate directory 'crd` for helm3.  So currently there is no good way to disable crds from base.  I have already created an issue before to track this 
https://github.com/istio/istio/issues/27609